assembly/src/assembler/callgraph.rs

```diff
@@ -0,0 +1,276 @@
+
+        // Make sure the callee is in the graph
+        self.get_or_insert_node(callee);
+        // Make sure the caller is in the graph
+        let callees = self.get_or_insert_node(caller);
+        // If the caller already references the callee, we're done
+        if callees.contains(&callee) {
+            return;
+        }
+
+        callees.push(callee);
+    }
+
+    /// Removes all edges to/from a procedure in `module`
+    ///
+    /// NOTE: If a procedure that is removed has predecessors (callers) in the graph, this will
+    /// remove those edges, and the graph will be incomplete and not reflect the "true" call graph.
+    /// In practice, we are recomputing the graph after making such modifications, so this a
+    /// temporary state of affairs - still, it is important to be aware of this behavior.
+    pub fn remove_edges_for_module(&mut self, module: ModuleIndex) {
+        for (_, out_edges) in self.nodes.iter_mut() {
+            out_edges.retain(|gid| gid.module != module);
+        }
+        self.nodes.retain(|gid, _| gid.module != module);
+    }
+
+    /// Removes the edge between `caller` and `callee` from the graph
+    pub fn remove_edge(&mut self, caller: GlobalProcedureIndex, callee: GlobalProcedureIndex) {
+        if let Some(out_edges) = self.nodes.get_mut(&caller) {
+            out_edges.retain(|n| *n != callee);
+        }
+    }
+
+    /// Returns the number of predecessors of `id` in the graph, i.e.
+    /// the number of procedures which call `id`.
+    pub fn num_predecessors(&self, id: GlobalProcedureIndex) -> usize {
+        self.nodes.iter().filter(|(_, out_edges)| out_edges.contains(&id)).count()
+    }
+
+    /// Construct the topological ordering of all nodes in the call graph.
+    ///
+    /// Returns `Err` if a cycle is detected in the graph
+    pub fn toposort(&self) -> Result<Vec<GlobalProcedureIndex>, CycleError> {
+        if self.nodes.is_empty() {
+            return Ok(vec![]);
+        }
+
+        let mut output = Vec::with_capacity(self.nodes.len());
+        let mut graph = self.clone();
+
+        // Build the set of roots by finding all nodes
+        // that have no predecessors
+        let mut has_preds = BTreeSet::default();
+        for (_node, out_edges) in graph.nodes.iter() {
+            for succ in out_edges.iter() {
+                has_preds.insert(*succ);
+            }
+        }
+        let mut roots =
+            VecDeque::from_iter(graph.nodes.keys().copied().filter(|n| !has_preds.contains(n)));
+
+        // If all nodes have predecessors, there must be a cycle, so
+        // just pick a node and let the algorithm find the cycle for
+        // that node so we have a useful error. Set a flag so that we
+        // can assert that the cycle was actually found as a sanity
+        // check
+        let mut expect_cycle = false;
+        if roots.is_empty() {
+            expect_cycle = true;
+            roots.extend(graph.nodes.keys().next());
```

Comment:
It's certainly possible to have multiple cycles simultaneously, depending on what has changed since the last compilation (if any). However detecting multiple cycles is not as simple as what is being done here, which relies on the simple property that failing to produce a topological sort means there is at least one cycle in the graph, but we don't actually try to work out the actual cycles. To actually identify all the cycles, we'd need to run something like Kosaraju's strongly-connected components algorithm, that identifies subgraphs that contain at least one cycle, but doesn't identify all of the cycles. Johnson's algorithm to find all of the elementary cycles of a graph is also an option, but is somewhat complex to implement.

TL;DR: we probably should hold off on implementing a fancier diagnostic here until we have a sense of how frequently this issue arises, and how often it involves multiple cycles vs accidentally introducing recursion in a new procedure. But if we feel there it is definitely worthwhile to do now, that's fine too.